### ISSUE TITLE
CNTKTextReader edge case problems

### ISSUE THREAD

**Issue by cmarschner** [Microsoft/CNTK#292]
1. Last line after last newline is ignored.
2. In case there was only one line without LF, it crashes with a division by 0. Same if file is empty.
[testconfig.txt](https://github.com/Microsoft/CNTK/files/190630/testconfig.txt)
[test.txt](https://github.com/Microsoft/CNTK/files/190631/test.txt)
3. Empty sequence causes crash.
[test-emptyseq.txt](https://github.com/Microsoft/CNTK/files/190634/test-emptyseq.txt)
4. Space instead of tab between sequence identifier and feature causes crash
[test-spc.txt](https://github.com/Microsoft/CNTK/files/190656/test-spc.txt)
5. Adding same input more than once causes debug assert, not error.
[test-multi.txt](https://github.com/Microsoft/CNTK/files/190658/test-multi.txt)

Expected:
- treat file as if it had a last LF (don't just ignore last line)
- handle empty files gracefully (warning)
- handle empty sequences gracefully (warning)
- handle empty minibatches (resulting from empty sequences) gracefully (warning)
- handle empty epochs (resulting from empty minibatches) gracefully (warning, nothing to do)

**Comment by frankseide** [comment]
I suggest some changes. We should clearly define the file format, and fail if it is malformed, while following the what-you-get-is-what-you-asked-for approach when it is not malformed. We should not try to infer user intent from malformed files. We should warn only very sparingly where it is much more likely that the user config is borked, than that the file is valid.

For 4., that means that if the last line has no LF, then the file is malformed and should be rejected as malformed. We should not hallucinate a LF.

This is very important. I have seen a situation where a final LF was silently imagined, and then users concatenated multiple files to form a larger corpus. Their files were plain text, so the last line without LF merged undetected with the first line of the concatenated corpus.

For 'Expected,' it means that we should not warn about an occasional empty sequences if they are a meaningful thing to allow. We should, however, keep warnings for large numbers of empty sequences (but no minibatches, because minibatches may contain only one sequence), which very very likely indicate that the user configuration is borked (or that some intermediate processing step caused empty sequences due to a config error or bad model).

**Comment by frankseide** [comment]
Same for TABs vs. spaces. If spaces are allowed, accept. If not, reject file as malformed, with an actionable error message. External file formats must always fully validated.

**Comment by raaaar** [comment]
Wrt 1, a quote from [wiki](https://github.com/Microsoft/CNTK/wiki/CNTKTextFormat-Reader):
`Each well-formed line must end with either a "Line Feed" \n or "Carriage Return, Line Feed" \r\n symbols (including the last line of the file).`

**Comment by alexeyo26** [comment]
> 4.Space instead of tab between sequence identifier and feature causes crash
This is not an expected behaviour. Bug?

**Comment by cmarschner** [comment]
While I agree with you in principle, what I mentioned is just the learnings from running experiments on TB scale for the past 8 years. 

Usually, scripts that generate these files are not well tested and have edge case problems. Often, they're one-offs as part of an experiment. They might be monolithic and rerunning them would cause significant resources and time. Out of my experience, some errors like this happen in almost _every_ nontrivial experiment.

Certainly what you don't want is the system to fail fast if one line in 200000 is wrong, which wouldn't have any effect on the training result. Especially for a case like (4), this would be a single error for a potentially huge file - no way the program should terminate for this one, IMHO. But it should tell me at least what was wrong and not just ignore the last line.

What has been implemented already in the reader is a "number of accepted errors" config option before it fails. I guess one can build upon that - if you want it to conform to the spec, set this count to 0. If you think the errors are benign, set it to a low number. If you want to get an overview of all the errors so that you can fix them in one swoop, set it to infinity. But what it entails is that at times a minibatch would be empty, which means in theory a whole epoch may be empty - ok maybe I went to far with this, it probably doesn't make sense to continue training in this case. I was perhaps thinking of an online training case where data is streamed - maybe that goes to far.

But I do think it is very important that in these cases we should always try to run _as long as possible_ so that you get as much feedback in one try.

**Comment by alexeyo26** [comment]
As a "man from the field and production" I would say that not well perfected scripts should be perfected over time. Because the alternative is all sorts of "permissive behaviour" when the reader "guesses" too much. This usually results in the debugging nightmare. And the nightmare will be especially "nightmarish" for huge data volumes.

**Comment by cmarschner** [comment]
Alright, I step back in my claim that the reader should "guess" - while I think it could easily be done in this particular case, there might not be too many other cases where this makes sense.

What I still feel quite strongly about, though, is that it should accept cases where data is missing due to badly formed input, which should just trickle up the chain so that the system can stay running while data is still read.

This is akin to compilers. While I agree with you that program source code should be "perfected over time" I do see value in a compiler not choking about the first unmatched bracket. Would you agree? 

I would also claim out of experience that in most cases the ROI of fixing a handful of issues in a large file is negative, so it would be irrational to do it.

**Comment by alexeyo26** [comment]
> I do see value in a compiler not choking about the first unmatched bracket

But this is exactly what ```maxErrors``` parameter is for.

**Comment by frankseide** [comment]
This is very risky if applied as a general principle. Machine learning is special in that little sloppinesses often go undetected and lead to little accuracy degradations. The sloppier you are, the more accuracy you are giving up. But often, we don't know what the final accuracy should be, so this goes undetected. That's why it is super-critical to pepper the code with consistency checks everywhere, to minimize the chance of an error going unnoticed. Second-guessing the user's intent has no place here.

Let us require to explicitly enable this error-tolerant behavior, *off* by default, and let us be explicit in what kinds of errors are tolerated. I suggest an explicit parameter `ignoreMalformedLines=true` where the name makes it very clear what is tolerated and what not. The default should be to fail on _any_ malformed input. `maxErrors` should only be used in conjunction with an explicit enabling option, and in any case should default to 0 (no errors allowed). Maybe rename to 'maxIgnoredLines'?

You still want to fix your scripts though. Speaking as someone who has run speech-recognition experiments for 22 years, trust me, sloppy corpus processing _will_ come back to bite you. It always does. Either in the form of undiscovered bugs causing minor accuracy degradation, or in the form of a long tail of downstream tools that all must know to tolerate the sloppy data file.

The example I alluded earlier to was really this: The speech trainer required two text files, one with pathnames and one with corresponding utterance transcriptions. Each one line per input file, and lines had to match between the two files. These text files were both created by concatenating such files from multiple corpora. One had been sloppily prepared and missed a LF at the end. As a consequence, in the concatenated file, the broken line was merged with the first line of the next file, and thus _the last 25% of the training corpus was off by 1_ between pathnames and their transcriptions. 25% of the data had invalid transcripts. Those led to bad alignments, which do happen during normal operation and are just skipped. As a consequence, the speech trainer silently tossed 25% of the training data! Final accuracy was only marginally worse, and it took a few weeks until this was discovered. Don't accept malformed data.

**Comment by cmarschner** [comment]
Ha ha, I think we funnel down on agreement (the example is an interesting one - if one error can cause this huge havoc...). Suffice it to say that these errors are sometimes out of one's control (we had a case where the cluster would create a malformed line every 200 MB, and this wasn't fixed by the cluster team for years. There was a workaround but it wasn't always used). 

So, to summarize:

1. Don't silently correct.
2. Don't accept errors by default. Report them and fail - already implemented
3. Allow for being lenient wrt. number of errors (with a config option) - already implemented

What you touched upon is:
4. Design the system such that the impact of an error is minized... - is this the case for ImageReader? Will it be the case for "composite readers"? E.g. there shouldn't be a dependency on the ordering of the file being perfect, the file being completely available etc. Rather, use some redundancy (extra metadata) to be able to discover these failures. We do have this in TextFormatReader with the explicit sequence ID, although right now this can be turned off in frame mode. That sequenceId needs to act as a key between readers, or there may be problems... I presume the ID will eventually have to be a string to be generic enough.

**Comment by frankseide** [comment]
The current plan for the sequence id was to make it a combination of a string and a number. This allows to use strings for everything, numbers for everything, or for forms of chunking ids (which may or may not coincide with chunking of data). Internally, they should always be mapped onto a numeric identifier (MBLayout currently uses a size_t).

**Comment by 1132520084** [comment]
In the wiki,
>Each separate line constitutes a sequence
>Each input can have only a single sample per line
For example, here is a SISO system(single input single output system) where X is the input and Y is the output. the data contain tow sequences, the input file is look approximately as follows:



	1 |X 0 |Y 1 |X 0.5 |Y 2 |X 1 |Y 3  
	2 |X 5 |Y 11 |X 5.5 |Y 12 |X 6 |Y 13

Is this right?
Can you give me an example?Thanks

**Comment by cmarschner** [comment]
@1132520084 Right now your format is not supported since one input can only be mentioned once per line. This is for performance reasons (maintenance of metadata). You have to split it up as follows:
1   |X 0 |Y 1
1   |X 0.5 |Y 2
1   |X 1 |Y 3  
2   |X 5 |Y 11
2   |X 5.5 |Y 12
2   |X 6 |Y 13
The sentence "each separate line constitutes a sequence" is not correct. Sequences are built a) per input that are b) in consecutive lines, merged by the sequence ID.

**Comment by 1132520084** [comment]
@cmarschner Thank you very much!
However , there are some errors in my computer. If i use the space as the delimiters between the sequences id and the input, the error is ` CNTK: Caught Win32 exception 0xc0000094: Integer division by zero `.

If I use the `tab(\t)` as the delimiters and the all the sequences id is different with each other, no problem. But if there are some lines' sequences id is same, an messagebox appear.

![capture](https://cloud.githubusercontent.com/assets/13234597/14349996/1f9bdc4e-fcf9-11e5-99ce-51a233de68d5.PNG)



And the error massage is `Assertion failed: reinterpret_cast<DenseSequenceData&>(*sample).m_numberOfSamples == 1, file SampleModePacker.cpp, line 108`.
This is the test file:
error
	
	1   |X 0 |Y 1
	1   |X 0.5 |Y 2
	1   |X 1 |Y 3  
	2   |X 5 |Y 11
	2   |X 5.5 |Y 12
	2   |X 6 |Y 13

no error


	1   |X 0 |Y 1
	2   |X 0.5 |Y 2
	3   |X 1 |Y 3  
	4   |X 5 |Y 11
	5   |X 5.5 |Y 12
	6   |X 6 |Y 13

**Comment by raaaar** [comment]
[8322296](https://github.com/Microsoft/CNTK/commit/8322296eabac5899df10d320fe9313169420b7de) addresses this issue. Please pull and give it another go.

**Comment by 1132520084** [comment]
@raaaar thank you.

**Comment by 1132520084** [comment]
@raaaar  a new error appears……
why the `TrainLossPerSample` is gradually becoming larger when I use the sequence id. Can you give me an simple example about sequence regression?
